scipy/stats/_hypotests.py

```diff
@@ -663,3 +664,333 @@ def somersd(x, y=None):
+    a 2x2 sample space using the `scipy.optimize.shgo` algorithm.
+
+    Parameters
+    ----------
+     total_c1, total_c2: int
+        The total sum of column 1 and column 2 in the initial variables array
+
+    index_arr: ndarray of boolean
+
+    n_iter: int
+        number of iteration inside shgo algorithm
+
+
+    Returns
+    -------
+    p_value : float
+        Return the maximum p-value considering every nuisance paramater
+        between 0 and 1
+
+    Notes
+    -----
+
+    Barnard exact test iterate over a nuisance parameter
+    :math:`\pi \in [0, 1]` to find the maximum p-value. To search this
+    maxima, this function uses `scipy.optimize.shgo`, which take
+    `n_iter` as iterations parameter.
+    Also, to compute the different combination used in the
+    p-values' computation formula, uses `gammaln` which is
+    more tolerant for large value than `scipy.special.comb`. This gives
+    a log combination. For the little precision lost, we gain a lot of
+    performance.
+    """
+    n = total_c1 + total_c2
+    x1 = np.arange(total_c1 + 1, dtype=np.int64).reshape(-1, 1, 1)
+    x2 = np.arange(total_c2 + 1, dtype=np.int64).reshape(1, -1, 1)
+    x1_sum_x2 = x1 + x2
+
+    x1_log_comb = _compute_log_combinations(total_c1)
+    x2_log_comb = _compute_log_combinations(total_c2)
+    x1_sum_x2_log_comb = x1_log_comb[x1] + x2_log_comb[x2]
+
+    def get_pvalue_from(nuisance_param):
+
+        with np.errstate(divide="ignore", invalid="ignore"):
+            log_nuisance = np.log(
+                nuisance_param,
+                out=np.zeros_like(nuisance_param),
+                where=nuisance_param >= 0,
+            )
+            log_1_minus_nuisance = np.log(
+                1 - nuisance_param,
+                out=np.zeros_like(nuisance_param),
+                where=1 - nuisance_param >= 0,
+            )
+
+            nuisance_power_x1_x2 = log_nuisance * (x1_sum_x2)
+            nuisance_power_x1_x2[(x1_sum_x2 == 0)[:, :, 0]] = 0
+
+            nuisance_power_n_minus_x1_x2 = log_1_minus_nuisance * (
+                n - x1_sum_x2
+            )
+            nuisance_power_n_minus_x1_x2[(x1_sum_x2 == n)[:, :, 0]] = 0
+
+            tmp_values_arr = np.exp(
+                x1_sum_x2_log_comb
+                + nuisance_power_x1_x2
+                + nuisance_power_n_minus_x1_x2
+            )
+
+        tmp_values_arr /= tmp_values_arr.sum(axis=(0, 1)).reshape(1, 1, -1)
+        # This operation compensate numerical errors because sums of
+        # p_values_arr should always be equal to one.
+
+        p_values_arr = tmp_values_arr[index_arr].sum(axis=0)
+        max_pvalue_index = p_values_arr.argmax()
+
+        p_value = p_values_arr[max_pvalue_index]  # take the max value
+        return p_value
+
+    # Since shgo find the minima, we need to take the negative value of the
+    # pvalue
+    result = shgo(
+        lambda x: -get_pvalue_from(x),
```

Comment:
If you have  a lambda here; do you believe that not making `get_pvalue_from` a closure anymore and using
```suggestion
        lambda x: -get_pvalue_from(x, x1_sum_x2, x1_sum_x2, x1_sum_x2_log_comb),
```

even puting the minus in the `get_minus_pvalue`, we can reduce to:

```suggestion
        lambda x: get_pvalue_from, x1_sum_x2, x1_sum_x2, x1_sum_x2_log_comb,
```
Should work as 

    The objective function to be minimized. Must be in the form f(x, *args), where x is
    the argument in the form of a 1-D array and args is a tuple of any additional fixed
    parameters needed to completely specify the function.

Would make it slightly clearer what the parameter of the function are ?
I'm going to guess this might also have a small performance effect as now you do not recreate a function on each call. 

With this  we get rid of any dynamic feature so could later be sent likely sent to `@numba.jit` or cythonized.
I'll let you judge if this is desirable or not, and or when to do that; it might be easier to keep it as a closure until deciding between grid search and  shgo.

Reply:
Thank you @Carreau, it it indeed a lot better. Changes are in 3775cf19e564c9e38c157d0745ac41824bfa5ad0